### ISSUE TITLE
[Q] How to start agent while using local sweep controller. 

### ISSUE THREAD

**Issue by MoH-assan** [wandb/wandb#8183]
I am running local sweep controller and agents on  [Niagra HPC](https://www.scinethpc.ca/niagara/)
On the login node, I successfully initialize the controller using
`$ wandb sweep --controller config.yaml --project 'wandb_offline_sweep' --entity 'uoft' `

On another terminal but on the same login node, I also successfully initialized an agent using 
`$ wandb agent  --count 2  uoft/wandb_offline_sweep/uwtwkm4i`

On a compute node (not connected to the internet), when I try 
`$ wandb agent  --count 2  uoft/wandb_offline_sweep/uwtwkm4i`
I get 
```
wandb: Starting wandb agent 🕵️
wandb: Network error (ConnectTimeout), entering retry loop.
```

I am not sure why Wandb is trying to connect to the internet. 

Note I tried,
`$ wandb controller uoft/wandb_offline_sweep/uwtwkm4i`
 and it fails both on the login and compute node. 

I am running minimal code from wandb documentation 

```
import wandb
import yaml
import random
import numpy as np
import os 
from pathlib import Path
os.environ['WANDB_MODE'] = 'offline'
os.environ["WANDB_INIT_TIMEOUT"] = "900"
os.environ["WANDB__SERVICE_WAIT"] = "900"
def train_one_epoch(epoch, lr, bs):
    acc = 0.25 + ((epoch / 30) + (random.random() / 10))
    loss = 0.2 + (1 - ((epoch - 1) / 10 + random.random() / 5))
    return acc, loss


def evaluate_one_epoch(epoch):
    acc = 0.1 + ((epoch / 20) + (random.random() / 10))
    loss = 0.25 + (1 - ((epoch - 1) / 10 + random.random() / 6))
    return acc, loss


def main():
    # Set up your default hyperparameters
    with open("./config.yaml") as file:
        config = yaml.load(file, Loader=yaml.FullLoader)
    
    run = wandb.init(config=config)
    run_id = run.id
    entity = run.entity
    project = run.project
    wandb_path = f"{entity}/{project}/{run_id}" 
    wandb_dir = Path(wandb.run.dir).parent.resolve()
    print(f"wandb_path: {wandb_path}")
    print(f'To sync the wandb run, run the following command \n: wandb sync {wandb_dir}',flush=True)   
    # Note that we define values from `wandb.config`
    # instead of  defining hard values
    lr = wandb.config.lr
    bs = wandb.config.batch_size
    epochs = wandb.config.epochs
    print(f'Running main with the hyperparameters: lr={lr}, bs={bs}, epochs={epochs}')
    for epoch in np.arange(1, epochs):
        train_acc, train_loss = train_one_epoch(epoch, lr, bs)
        val_acc, val_loss = evaluate_one_epoch(epoch)

        wandb.log(
            {
                "epoch": epoch,
                "train_acc": train_acc,
                "train_loss": train_loss,
                "val_acc": val_acc,
                "val_loss": val_loss,
            }
        )


# Call the main function.
main()
```

with the following config 

```
# config.yaml
program: /gpfs/fs0/scratch/c/cglee/abubakr5/Misc/wb_sweep/main.py
method: random
controller:
  type: local
name: sweep
metric:
  goal: maximize
  name: val_acc
parameters:
  batch_size: 
    values: [16,32,64]
  lr:
    min: 0.0001
    max: 0.1
  epochs:
    values: [5, 10, 15]
```

**Comment by fmamberti-wandb** [comment]
Hi @MoH-assan , thank you for reaching out with your question.

When setting up the local controller on the login node, this will only be available for agents on the login node as it's only available locally.

You would need to setup a local controller on the compute node to run an agent on it however, this is currently not possible as access to W&B cloud is required to setup a local controller. 

Currently running W&B Sweeps in offline mode for hp tuning is not supported - would you like me to raise this as a feature request on your behalf? If so, would you mind sharing additional information on your use case so I can pass it to our product team to review?

**Comment by MoH-assan** [comment]
@fmamberti-wandb  
> would you like me to raise this as a feature request on your behalf? 

This would be great. 

> would you mind sharing additional information on your use case so I can pass it to our product team to review?

Currently,
I using W&B while doing hyperparameter tuning for the DL model. Currently, I am doing a manual grid search. I use Slurm to submit individual jobs, one per each hyper-parameter configuration. These jobs are executed on different compute nodes that have no access to the Internet but share the same file system with login nodes.
Note the login nodes do have access to the Internet. I can ssh the compute node from the login node once the compute node is allocated to me. 
I use [wandb-osh](https://github.com/klieret/wandb-offline-sync-hook) for uploading the to W&B  server, where I later process and visualize the hyper-tunning result either online or using W&B API. 

I would like to
Use W&B sweep to automate the hyperparameter and take advantage of Bayesian optimization. Ideally, I would like to have the controller running on the login node with agents distributed across the compute nodes.

I am not sure if these solutions address my needs, but I will be looking into them. I would appreciate your thoughts and would appreciate it if any of them could be integrated into W&B. 

- [WandB-on-HPC](https://github.com/ptzpanzer/WandB-on-HPC)
- [wandb_on_slurm](https://github.com/elyall/wandb_on_slurm)
- This  [ W&B blog ](https://wandb.ai/events/SLURM/reports/Improving-Your-Deep-Learning-Workflows-with-SLURM-and-Weights-Biases--VmlldzozOTU4NzU0) closing statement provides a number of suggestions.

**Comment by fmamberti-wandb** [comment]
Hi @MoH-assan, thanks for sharing these details on your request. I have raised this with our product team to review together with the repositories you suggested and we will keep you posted.

Regarding [WandB-on-HPC](https://github.com/ptzpanzer/WandB-on-HPC), it seems they are trying to solve the problem you are facing however, the latest updates are ~1 year old and wouldn't be something currently supported by `wandb` so there is no guarantee that this would work now or in the future.
